waiter/src/waiter/service_description.clj

```diff
@@ -738,6 +741,10 @@
             (validate service-description-builder service-description {:allow-missing-required-fields? false}))
           {:core-service-description core-service-description
            :on-the-fly? contains-waiter-header?
+           :retrieve-fallback-service-description (fn []
```

Comment:
What are the merits of this approach over repeated calls to `compute-service-description` directly?

Reply:
Is your argument to inline the recursive call or to require the caller to make the explicit call to `compute-service-description`? I assume the latter as the first approach is a trivial change (and a matter of taste).

Remember that the user is actually calling `request->descriptor`. We could choose to offer a new function `descriptor->fallback-descriptor` or use this new method. With this approach we are generating and sending the function isntead of requiring the user to know about and invoke `descriptor->fallback-descriptor` directly. Also, some arguments such as `sources`, `waiter-headers`, `passthrough-headers`, ...) are either generated internally from other calls. With this approach, we do not require regeneration of the parameters (i.e. saving re-computation costs).

Reply:
> In your example, does request->descriptor gain a new dependency on the state of running services?

It would be the job of `wrap-descriptor` or another filter that runs after `wrap-descriptor` to correct the descriptor based on running services. `request->descriptor` would remain as it is now, returning a descriptor based on the current request.